obspy/signal/tests/test_cross_correlation.py

```diff
@@ -271,6 +271,81 @@ def test_xcorr_pick_correction_images(self):
 
+    def test_template_matching(self):
+        """
+        Test for moving window correlations with "full" normalisation.
+
+        Comparison result is from EQcorrscan v.0.2.7.
+        """
+        result = [
+            -2.24548906e-01,  7.10350871e-02,  2.68642932e-01,  2.75941312e-01,
+            1.66854098e-01,  1.66086946e-02, -1.29057273e-01, -1.96172655e-01,
+            -1.41613603e-01, -6.83271606e-03,  1.45768464e-01,  2.42143899e-01,
+            1.98310092e-01,  2.16377302e-04, -2.41576880e-01, -4.00586188e-01,
+            -4.32240069e-01, -2.88735539e-01,  1.26461715e-01,  7.09268868e-01,
+            9.99999940e-01,  7.22769439e-01,  1.75955653e-01, -2.46459037e-01,
+            -4.34027880e-01, -4.32590246e-01, -2.67131507e-01, -6.78363896e-04,
+            2.08171085e-01,  2.32197508e-01,  8.64804164e-02, -1.14158235e-01,
+            -2.53621429e-01, -2.62945205e-01, -1.40505865e-01,  3.35594788e-02,
+            1.77415669e-01,  2.72263527e-01,  2.81718552e-01,  1.38080209e-01,
+            -1.27307668e-01]
+        data = read()[0].data
+        template = data[400:600]
+        data = data[380:620]
+        shift = 20
+        cc = correlate(data, template, shift, normalize="full", domain="freq")
+        np.testing.assert_allclose(cc, result, atol=0.01)
+        shift, corr = xcorr_max(cc)
+        self.assertAlmostEqual(corr, 1.0, 2)
+        self.assertEqual(shift, 0)
+
+    def test_template_matching_time(self):
+        result = [
+            -2.24548906e-01,  7.10350871e-02,  2.68642932e-01,  2.75941312e-01,
+            1.66854098e-01,  1.66086946e-02, -1.29057273e-01, -1.96172655e-01,
+            -1.41613603e-01, -6.83271606e-03,  1.45768464e-01,  2.42143899e-01,
+            1.98310092e-01,  2.16377302e-04, -2.41576880e-01, -4.00586188e-01,
+            -4.32240069e-01, -2.88735539e-01,  1.26461715e-01,  7.09268868e-01,
+            9.99999940e-01,  7.22769439e-01,  1.75955653e-01, -2.46459037e-01,
+            -4.34027880e-01, -4.32590246e-01, -2.67131507e-01, -6.78363896e-04,
+            2.08171085e-01,  2.32197508e-01,  8.64804164e-02, -1.14158235e-01,
+            -2.53621429e-01, -2.62945205e-01, -1.40505865e-01,  3.35594788e-02,
+            1.77415669e-01,  2.72263527e-01,  2.81718552e-01,  1.38080209e-01,
+            -1.27307668e-01]
+        data = read()[0].data
+        template = data[400:600]
+        data = data[380:620]
+        shift = 20
+        cc = correlate(data, template, shift, normalize="full", domain="time")
+        np.testing.assert_allclose(cc, result, atol=0.01)
+        shift, corr = xcorr_max(cc)
+        self.assertAlmostEqual(corr, 1.0, 2)
+        self.assertEqual(shift, 0)
+
+    """
+    We need a test that does this, but the only way to do it with long duration
+    data - EQcorrscan tests against Kaikoura data, which we know causes issues
+    for a lot of cross-correlation implementations.
+    """
```

Comment:
How big is the data?And could you not just use some random noise with a large spike in the middle?

Reply:
In all honesty, I have tried with random data with spikes, but I haven't been able to usefully reproduce the amplitude ranges in the data, so gave up.  With float 64 in C the issues only occur after about 16 hours of data during the sequence (sampled at 100Hz), so it is quite a chunk of data required to get this issue. Unfortunately these issues occur exactly when (in one case) you want correlations to be robust: for the detection of early aftershocks or triggered seismicity.
The dataset that I use in previous tests is multi-channel (6-channels, 8640001 samples each) and is 474MB on disk as MiniSeed - looking at top it looks like that inflates to about 1.3GB in RAM.
Testing that locally suggests that we aren't running into floating point rounding errors here.

Reply:
Where does the instability come from? To my this line:

```python
denominator = np.cumsum(long_data ** 2)
```

seems to be a candidate but even with data still in counts and very long time series and a short template I have a hard time imagining this getting into a regime where double precision floating point arithmetics gets inaccurate enough to notice.

The whole problem (if its due to this line) could be mitigated by not using a single cumulative sum but many of them with some special handling of the boundary regions - not sure if its worth the effort to be honest.

Reply:
Some short test data that might work: Use some data you know and prepend some very high amplitude random data to it. Then do the CC with a fairly short template and only compare the result for the later part to a known good one. This should "preload" the cumulative sum so rounding errors might become noticeable. To test if this is actually a problematic data set you could run it against a CC implementation that you know has problems with these kind of things.

```python
data = np.concatenate([np.random.random(10) * 1e200, obspy.read()[0].data])
```

Reply:
That is exactly the line that would cause the instability - and you are right, one way to mitigate would be to break the data down, but I'm not sure it is worth the effort either. It doesn't seem to be an issue with the Kaikoura data.
By the way, I agree with the issue being less likely with data in counts, but often in template matching the data are filtered before correlation, yielding data in floats.